cmd/main.go

```diff
@@ -59,6 +73,129 @@ func getPlayer(sa *spotify.Spotify) echo.HandlerFunc {
+			// This will more than likely happen in the case where nothing is playing or authentication
+			// fails
+			utils.Logger.Error("Error getting playerstate", zap.Error(err))
+		}
+
+		// Check if player is not nil and that the player has an item
+		if player != nil && player.Item != nil {
+
+			// Create a track from the playerstate
+			track := makeTrack(player)
+
+			// This function requires data from the previous loop so it needs to be called before the update to the playerstate
+			// This is to check if the track has changed and if so add it to the db or if the track has been replayed
+			dbCheckUpdate(dbClient, track, player.Progress, ctx)
+
+			// This function updates the playerstate with the new track and progress 
+			playerState = updatePlayerState(track, player.Progress)
+		}
+
+		// For testing to see if the loop is working
+		//utils.Logger.Info("Playerstate recieved", zap.Any("player", playerState))
+		// Debugging Query to see if the tracks are being added to the db correctly 
+		// Best to use len since it removes some of the clutter from the log
+		tr, err := dbClient.Track.Query().All(ctx)
+		if err != nil {
+			utils.Logger.Error("Error querying tracks", zap.Error(err))
+		}
+		utils.Logger.Info("Tracks", zap.Any("tracks", len(tr)))
+
+		time.Sleep(5 * time.Second)
+	}
+}
+
+func makeTrack(player *spotifyLib.PlayerState) *ent.Track {
+	// Check if playerState is active if not return previous track as fallback
+
+	artists := make([]string, len(player.Item.Artists))
+	//TODO: For some reason the genres field is not available
+	// From Item.Artists despite the API saying so. This may be connected
+	// to a type issue where Item is not locked to being a track object
+	// genres := make(map[string]byte)
+	// e.g. player.Item.SimpleTrack.Artists[0].Genres
+
+	for i, artist := range player.Item.Artists {
+		artists[i] = artist.Name
+	}
+	track := &ent.Track{
+		Name:          player.Item.Name,
+		Artists:       artists,
+		ArtistsGenres: nil,
+		AlbumName:     player.Item.Album.Name,
+		AlbumImageURI: player.Item.Album.Images[0].URL,
+		DurationMs:    int32(player.Item.Duration),
+		URI:           string(player.Item.URI),
+	}
+
+	// Return a pointer to a track
+	return track
+
+}
+
+func updatePlayerState(track *ent.Track, progress int) *PlayerState {
+	if track.Name == "" {
+		return playerState
+	}
+
+	//TODO: not sure if this is how you do pointer updates in Golang
+	ps := &PlayerState{*track, progress, time.Now()}
+	return ps
+}
+
+func dbCheckUpdate(dbClient *ent.Client, track *ent.Track, progress int, ctx context.Context) {
+
+	// Check if the track has just changed and if so add it to the db	
+	if playerState.Track.Name != track.Name {
+		addTrack(dbClient, track, ctx)
+	}
+
+	// Check for replays
+	// TODO: Maybe find a better way to do this but works for now
+	if (int(track.DurationMs/50)*100) < playerState.Progress && progress <= int((track.DurationMs/05)*100) {
+		addTrack(dbClient, track, ctx)
+	}
+
+}
```

Comment:
We should write tests for logic like this :)

Reply:
Is it alright if I do this on another branch so that I don't end up blocking you? I'll write an issue about it :)

Reply:
Of course.

Apparently i cant comment without it being a review comment, so just close this conversation when the issue has been created